### DISCUSSION TITLE
Umami, Controller Input and more for VR Chat.

### DISCUSSION THREAD

**Discussion by alatnet** [vrchat-community/osc#118] [Show and tell]
Hello! I’ve created an application for end users and world creators/item creators to add controller support, such as Flight Sticks or Racing Wheels, to their worlds or items.
It allows end users to easily remap their controllers for near limitless controller configuration for worlds and items.
Features to come is Controller Profile support and Auto Profile Switching.

Hope this gives creators more access to inputs than before.

ask.vrchat.com page: https://ask.vrchat.com/t/umami-controller-input-and-more-for-vr-chat/8834
Github releases: https://github.com/alatnet/Umami/releases

**Comment by anonmon** [comment]
This is just for worlds, and items (in worlds? Not entirely sure what "Items" entails in a VRChat context.)
Not gonna lie, this seems like it'd be infinitely more useful if it also allowed for easily sending/receiving OSC to avatars as well, as there's no equivalent for doing that yet as far as I've found.

**Reply by alatnet** [reply]
Well, it's intended for using traditional and sim specific controllers such as a ps3/4 gamepad or flight sticks for things like vehicles (like the ones in the Flight Test world).  But, it can be used for avatars if you wish.  The addresses used are listed in the github so they can be used with anything.  The context for "items" can be used for either vehicles or for any "item" that a world/item creator wish to have controllable via game controllers.

EDIT: wait, do you mean this? https://docs.vrchat.com/docs/osc-as-input-controller

**Reply by anonmon** [reply]
Ah, I didn't read the readme thoroughly enough? Went looking for where you'd set names for parameters / address to then configure how controllers would be mapped to those names/addresses, but by the look of it you're supposed to set parameter names based on what Umami wants? I guess since it was made with worlds / vehicles in worlds in mind it wasn't set up to map out avatar parameters (Which go "/avatar/parameters/(ParameterName)" _specifically_, so the "/Umami/input/(Axis/Button/POV) wouldn't really work.")

I don't mean using OSC as a controller. I was thinking something along the lines of making glove controllers, or any sort of physical buttons or analog inputs alongside by way of a microcontroller, and having that map to a specific avatar's functionality, like more axis of control on fingers and such. And doing that with wired microcontrollers acting as DirectInput/Xinput devices would be infinitely easier than either having to write up dedicated OSC software on the PC side, or running it over WiFi like with the OSCLib-for-ESP8266 which would be a touch ridiculous for certain projects. Plus would make the hypothetical controller far more portable to other games and for using with other things in SteamVR (OVR Advanced Settings for instance) if it was already emulating a gamepad to work.

**Comment by alatnet** [comment]
0.6.0-beta released! Added a controller profile feature to be able to save your controller configuration.
